compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1545,23 +1545,31 @@ private boolean isAssignableToUnionType(BType source, BType target, List<TypePai
         return !notAssignable;
     }
 
-    public boolean isAssignableToFiniteType(BType type,
-                                            BLangLiteral literalExpr) {
-        if (type.tag == TypeTags.FINITE) {
-            BFiniteType expType = (BFiniteType) type;
-            boolean foundMember = expType.valueSpace
-                    .stream()
-                    .anyMatch(memberLiteral -> {
-                        if (((BLangLiteral) memberLiteral).value == null) {
-                            return literalExpr.value == null;
-                        } else {
-                            return (((BLangLiteral) memberLiteral).value.equals(literalExpr.value) &&
-                                    ((BLangLiteral) memberLiteral).typeTag == literalExpr.typeTag);
-                        }
-                    });
-            return foundMember;
+    boolean isAssignableToFiniteType(BType type, BLangLiteral literalExpr) {
+        if (type.tag != TypeTags.FINITE) {
+            return false;
         }
-        return false;
+
+        BFiniteType expType = (BFiniteType) type;
+        long matchCount = expType.valueSpace.stream().filter(memberLiteral -> {
+            if (((BLangLiteral) memberLiteral).value == null) {
+                return literalExpr.value == null;
+            }
+            // Check whether the member literal and the literal that needs to be tested are of same kind and check
+            // the value equality between them.
+            return isSameLiteralKind((BLangLiteral) memberLiteral, literalExpr) &&
+                    ((BLangLiteral) memberLiteral).value.equals(literalExpr.value);
+        }).count();
+
+        // If more than one match means the value space contains ambiguous values.
+        if (matchCount > 1) {
+            dlog.error(literalExpr.pos, DiagnosticCode.AMBIGUOUS_TYPES, type);
```

Comment:
Shall we add some test cases?

Reply:
This needs to be revisited as per the comments of `James` in the mail thread "Clarification regarding union types" in Ballerina-dev group. Basically, we should not treat this scenario as ambiguous. There will be some other logic changes as well, which I am planning to handle in a different PR. I will add test cases when I revisit this logic for the fulfillness.

Reply:
lets add a test(s) for the current behavior. We can alter it later once we do the improvements